solutions/usaco-531.mdx

```diff
@@ -0,0 +1,124 @@
+---
+id: usaco-531
+source: USACO Silver 2015 February
+title: Superbull
+author: Maggie Liu
+---
+
+[Official Analysis](http://www.usaco.org/current/data/sol_superbull_silver.html)
+
+## Explanation
+
+We can use Kruskal's to find the maximum spanning tree of the graph formed by
+the teams. First, add all possible edges between two teams to the vector
+$\texttt{edges}$ and sort it in decreasing order of edge weight. For each edge,
+if we can unite the connected components of the two vertices, then we have
+eliminated one team. Increment $\texttt{eliminated}$ by $1$ and add the edge
+weight to our answer, $\texttt{ans}$. Once $n - 1$ teams have been eliminated,
+we can print $\texttt{ans}$.
+
+## Implementation
+
+**Time Complexity:** $\mathcal{O}(N^2 \log N)$
+
+<LanguageSection>
+
+<CPPSection>
+
+```cpp
+#include <iostream>
+#include <cstdio>
+#include <algorithm>
+#include <vector>
+using namespace std;
+void init(int n);
+bool unite(int a, int b);
+
+const int MAX_N = 2000;
+int parent[MAX_N];
+int compsize[MAX_N];
+struct Edge
+{
+	int from, to, weight;
+};
+int main()
+{
+	freopen("superbull.in", "r", stdin);
+	freopen("superbull.out", "w", stdout);
+	int n;
+	cin >> n;
+	int ids[MAX_N];
+	for (int i = 0; i < n; i++)
+	{
+		cin >> ids[i];
+	}
+	vector<Edge> edges;
+	for (int i = 0; i < n; i++)
+	{
+		for (int j = i + 1; j < n; j++)
+		{
+			edges.push_back({i, j, ids[i] ^ ids[j]});
+		}
+	}
+	init(n);
+	// sort edges in decreasing order of weight
+	sort(edges.begin(), edges.end(), [](Edge a, Edge b){
+		return a.weight > b.weight;
+	});
+	int eliminated = 0;
+	long long ans = 0;
+	for (Edge e : edges)
+	{
+		if (unite(e.from, e.to))
+		{
+			eliminated++;
+			ans += e.weight;
+			// if all but 1 team has been eliminated
+			if (eliminated == n - 1)
```

Comment:
[optional] idt you need to check whether `eliminated == n - 1` (just print `ans` at the end) because there will be exactly `n - 1` DSU merge operations (think of it as a tree).

Reply:
isn't it slightly faster since you don't have to finish looping through all the other edges?

Reply:
oh yeah thats true :+1: